### PR TITLE
Frontend and tests for directory

### PR DESCRIPTION
**Pull request by sezaljain**

### CODE REVIEW

File: src/cljs/nnts2/directory/events.cljs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+(ns nnts2.directory.events
+  (:require [re-frame.core :as re-frame]
+            [nnts2.directory.api-data :as api-data]))
+
+(enable-console-print!)
+
+(re-frame/reg-event-fx
+ ::create-directory-submit
+ (fn [cofx [_ dir-data]]
+   (prn dir-data)
```
Comment (nid90) [inline src/cljs/nnts2/directory/events.cljs]: Stray `prn`.

**Comment by kitallis** [comment]
I think overall, it's best to keep related changes in separate PRs. For example, UI can go as a separate change, specs can go either in the backend PR or another one.

File: src/cljs/nnts2/organization/events.cljs [hunk 2]
```diff
@@ -14,14 +15,18 @@
...
+       (assoc-in [:organization (:id org-details)] (dissoc org-details :id))
        (assoc-in [:organization :show-create-org-form] false))))
 
-(re-frame/reg-event-db
+(re-frame/reg-event-fx
  ::organizations-retrieved
- (fn [db [_ org-details]]
-   (-> db
-       (assoc-in [:organization :data :all] org-details))))
+ (fn [cofx [_ org-details]]
```
Comment (kitallis) [inline src/cljs/nnts2/organization/events.cljs]: What does `cofx` represent?

File: src/cljs/nnts2/organization/events.cljs [hunk 3]
```diff
@@ -14,14 +15,18 @@
...
+   {:http-xhrio (map #(dir-api/get-directories (:id %)) org-details)
+    :db (assoc (:db cofx) :organization
+               (reduce
+                (fn [acc v] (assoc acc (:id v) (dissoc v :id)))
```
Comment (kitallis) [inline src/cljs/nnts2/organization/events.cljs]: I think a `dissoc` here is not necessary, it just makes the `reduce` more noisy.

File: src/cljs/nnts2/organization/views.cljs [hunk 4]
```diff
@@ -50,4 +56,5 @@
          (if @show-create-org-form
            [create-form]
            [:div])
-         (map sidenav-item orgs)]))))
+         (for [[k v] orgs]
+           (if k [organization-view k v]))]))))
```
Comment (kitallis) [inline src/cljs/nnts2/organization/views.cljs:58]: It seems like extracting the `id` from the map is a bit redundant here. Why can't we simply pass the list of orgs over and destructure the `id` out with the remaining `name` for `organization-view`. That way creating a map earlier in `reg-event-fx` will not really be necessary.

File: test/clj/nnts2/directory/db_test.clj [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  (testing "should create dir in org when given existing user, org and dir name"
+    (let [fake-dir-data  (factory/directory "fake-dir")
+          output (directory-db/create fake-dir-data)]
+      (is (not (empty? output)))
+      (is (= fake-dir-data(select-keys output [:name :parent-id :org-id :created-by-id]))))))
+
+
+(deftest create-nested-dir-test
+  (testing "should create sub directory in directory when given name, existing directory,userwith correct org"
+    (let [fake-dir-data (factory/directory "fakedir")
```
Comment (kitallis) [inline test/clj/nnts2/directory/db_test.clj]: I think we can avoid calling them `fake-` repeatedly. It's in the test namespace, it's mostly apparent that it's fake.

File: src/cljs/nnts2/directory/subs.cljs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+ ::org-directories
+ (fn [db [_ org-id]]
+   (get-in db [:organization org-id :directories])))
+
+(re-frame/reg-sub
+ ::selected-directory
+ (fn [db [_ dir-id]]
+   (if dir-id
+     (if (= dir-id (:selected-dir db)) true false)
+     false)))
```
Comment (kitallis) [inline src/cljs/nnts2/directory/subs.cljs]: What case is when the selected directory is actually `nil`?

Also, this `if-else` ladder can be re-written more simply as,

```
(and dir-id
     (= dir-id (:selected-dir db)))
```

File: src/cljs/nnts2/directory/components.cljs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  (let [selected (re-frame/subscribe [::subs/selected-directory (:id dir)])
+        add-new (r/atom false)]
+    (fn []
+      [:text
+       {:id (:id dir)
+        :onDoubleClick (fn [e] (swap! add-new not))
+        :on-click #(re-frame/dispatch [:nnts2.directory.events/directory-selected (:id dir)])
+        :style (if @selected {:color "orange"} {})}
+       (:name dir)
+       [:dl  [:dt (if (not@add-new) {:style  {:display "none"}}) [add-child-directory dir]]]])))
```
Comment (kitallis) [inline src/cljs/nnts2/directory/components.cljs]: What is `not@add-new`? How does that even work? Is it missing a space?

File: src/cljs/nnts2/directory/events.cljs (hunk 1 above)
Comment (kitallis) [inline src/cljs/nnts2/directory/events.cljs]: Stray `prn`.

File: src/clj/nnts2/model/directory.clj [hunk 8]
```diff
@@ -38,4 +41,11 @@
 
 
 (defn create [params]
-  (res/response (db/create params)))
+  " create directory if org exists and orgs match if there is a parent directory mentioned"
+  (if (org/org-exists (:org-id params) (:created-by-id params))
+    (if (nil? (:parent-id params))
```
Comment (kitallis) [inline src/clj/nnts2/model/directory.clj]: I think `nil?` is unnecessary here. `nil` is a falsey value.

File: src/clj/nnts2/model/directory.clj [hunk 9]
```diff
@@ -16,10 +17,12 @@
         next-iter-rows (clojure.set/difference (set remaining-rows) (set to-insert))]
     (reduce
      (fn [acc v]
```
Comment (kitallis) [inline src/clj/nnts2/model/directory.clj]: Can we rename the `v` to be the element type in question? Like "directory"?

Can destructure: `id, name, org-id` in the function param itself.

**Review by kitallis — CHANGES_REQUESTED** [review]

File: src/cljs/nnts2/directory/components.cljs (hunk 7 above)
Comment (sezaljain) [reply src/cljs/nnts2/directory/components.cljs]: its missing a space. will find out how it even compiled.

File: src/cljs/nnts2/directory/subs.cljs (hunk 6 above)
Comment (sezaljain) [reply src/cljs/nnts2/directory/subs.cljs]: if you are clicking on organizations.. then dir-id is null.
I am treating org and directory in the same way on the sidebar. That means on-click same event is called, they are being rendered using the same view.
The difference between the two is this dir-id = nil

File: test/clj/nnts2/directory/db_test.clj (hunk 5 above)
Comment (sezaljain) [reply test/clj/nnts2/directory/db_test.clj]: yes, will do. it was annoying me also. But I wrote it once so kept doing that. Will change

File: src/cljs/nnts2/organization/events.cljs (hunk 2 above)
Comment (sezaljain) [reply src/cljs/nnts2/organization/events.cljs]: coeffects.
This is passed in along with event as arguments to the reg_event_fx handler.

**Comment by sezaljain** [comment]
> I think overall, it's best to keep related changes in separate PRs. For example, UI can go as a separate change, specs can go either in the backend PR or another one.

File: src/clj/nnts2/model/directory.clj (hunk 8 above)
Comment (sezaljain) [reply src/clj/nnts2/model/directory.clj]: this is not clear. to be discussed

File: src/cljs/nnts2/organization/views.cljs (hunk 4 above)
Comment (sezaljain) [reply src/cljs/nnts2/organization/views.cljs:58]: This I need to discuss with you.
This is all becuase of the way I am storing data in frontend.
The db map looks something like
```
{:organization
 {:id1
  {:name "org-name1"
   :directories [{:id "dir-id1", :name "test-dir1"}]}}}
```

**Pull request merged**